src/vulkan_state.cpp

```diff
@@ -21,29 +21,63 @@
  */
 
 #include "vulkan_state.h"
-#include "vulkan_wsi.h"
-
+#include "device_uuid.h"
 #include "log.h"
 
+#include <array>
 #include <vector>
-#include <algorithm>
+#include <vulkan/vulkan.hpp>
+
 
-VulkanState::VulkanState(VulkanWSI& vulkan_wsi)
+VulkanState::VulkanState(VulkanWSI& vulkan_wsi, ChoosePhysicalDeviceStrategy const& pd_strategy)
 {
     create_instance(vulkan_wsi);
-    choose_physical_device(vulkan_wsi);
-    create_device(vulkan_wsi);
+    create_physical_device(vulkan_wsi, pd_strategy);
+    create_logical_device(vulkan_wsi);
     create_command_pool();
 }
 
-void VulkanState::log_info()
+std::vector<vk::PhysicalDevice> VulkanState::available_devices(VulkanWSI& vulkan_wsi) const
 {
-    auto const props = physical_device().getProperties();
+    auto available_devices = instance().enumeratePhysicalDevices();
+    for (auto it_device = available_devices.begin(); it_device < available_devices.end(); ++it_device)
+    {
+        if (!vulkan_wsi.is_physical_device_supported(*it_device))
+        {
+            available_devices.erase(it_device);
```

Comment:
The std::vector::erase() method invalidates iterators at or after the point of the erase, so the above is problematic. The safe pattern in this case is:

```c++
for (auto iter = v.iter(); iter != v.end(); )
{
    if (...)
    {
        ... use iter ...
        iter = v.erase(iter);
    }
    else
    {
        ++iter;
    }
}

Reply:
I didn't know. Thank you for sharing your knowledge